[![Build Status](https://travis-ci.org/Oshlack/STRetch.svg?branch=master)](https://travis-ci.org/Oshlack/STRetch)

[![codecov](https://codecov.io/github/Oshlack/STRetch/branch/master/graphs/badge.svg)](https://codecov.io/github/Oshlack/STRetch)


## [STRetch user guide](https://github.com/Oshlack/STRetch/wiki)

**Update:** the STRetch paper is now published!

If using STRetch, please cite:

Dashnow, Harriet, Monkol Lek, Belinda Phipson, Andreas Halman, Simon Sadedin, Andrew Lonsdale, Mark Davis, et al. 2018. “STRetch: Detecting and Discovering Pathogenic Short Tandem Repeat Expansions.” Genome Biology 19 (1): 121. [https://doi.org/10.1186/s13059-018-1505-2](https://doi.org/10.1186/s13059-018-1505-2)

# STRetch

Method for detecting short tandem repeat expansions from short-read sequencing data.

STRetch requires paired end reads to work. It is primarily intended for use on
whole genome sequencing data, but can also be run on exome or other targeted
sequencing data, although size estimates should be interpreted with caution.

Please see the [STRetch user guide](https://github.com/Oshlack/STRetch/wiki)
for installation and usage instructions.
